### PR TITLE
[ADP-2788] add max-count to listTransactions API call

### PR DESCRIPTION
**Pull request by paolino**
-- add an optional max_count parameter to the listTransactions API call

ADP-2788

### CODE REVIEW

File: lib/wallet/api/http/Cardano/CLI.hs [hunk 1]
```diff
@@ -1353,6 +1357,14 @@ sortOrderOption = optionT $ mempty
     <> help "specifies a sort order, either 'ascending' or 'descending'."
     <> showDefaultWith showT
 
+-- | [--order=ORDER]
```
Comment (piotr-iohk) [inline lib/wallet/api/http/Cardano/CLI.hs]:
```suggestion
-- | [--max_count=MAX_COUNT]
```

File: lib/wallet/integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs [hunk 2]
```diff
@@ -1812,8 +1824,18 @@ spec = describe "SHELLEY_TRANSACTIONS" $ do
...
+                Nothing
               length <$> [txs1, txs2] `shouldSatisfy` all (== 0)
+    
+    it "TRANS_LIST_LIMIT_01 - \
+       \Transactions can be limited" $
+          \ctx -> runResourceT $ do
+              w <- fixtureWalletWith @n ctx 
+                    (replicate 100 $ minUTxOValue (_mainEra ctx))
+              txs <- listLimitedTransactions @n ctx w 9
+              length txs `shouldBe` 9
```
Comment (piotr-iohk) [inline lib/wallet/integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs:1859]: I think we need more detailed testing, in particular showing how `max_count` plays with `order`, `start` and `end` query parameters. Expectation would be that `max_count` respects constraints of other parameters as well. It is also not clear how `max_count` should behave together with `order`. Possible scenarios:

- `max_count = n` => n most recent txs in descending order
- `max_count = n, order = descending` => n most recent txs in descending order
- `max_count = n, order = ascending` => n most recent txs in ascending order **or** n oldest transactions in ascending order (requirements are not percise)
- `max_count = n, start = date` => n most recent txs from start date in descending order
- `max_count = n, start = date, order = descending` => n most recent txs from start date in descending order
-  `max_count = n, start = date, order = ascending` => n most recent txs from start date in ascending order **or** n oldest transactions starting from start date in ascending order (requirements are not percise)
- `max_count = n, end = date` => n most recent txs not older than end date in descending order
- `max_count = n, end = date , order = descending` => n most recent txs not older than end date in descending order
- `max_count = n, end = date , order = ascending` => n most recent txs not older than end date in descending order **or** n oldest transactions not older than end date in ascending order (requirements are not percise)
-  `max_count = n, start = date1, end=date2` => n most recent txs from start date, not older than end date in descending order
- `max_count = n, start = date1, end=date2, order=descending` => n most recent txs from start date, not older than end date in descending order
- `max_count = n, start = date1, end=date2, order=ascending` => n most recent txs from start date, not older than end date in ascending order **or** n oldest transactions starting from start date and not older than end date in ascending order (requirements are not percise)

File: lib/wallet/integration/src/Test/Integration/Framework/DSL.hs [hunk 3]
```diff
@@ -2575,7 +2576,21 @@ listAllTransactions
...
+        ( DecodeAddress n
+        , DecodeStakeAddress n
+        , HasType (ApiT WalletId) w
+        , MonadUnliftIO m
+        )
+    => Context
+    -> w
+    -> Int
+    -> m [ApiTransaction n]
+listLimitedTransactions ctx w limit = do
```
Comment (jonathanknowles) [inline lib/wallet/integration/src/Test/Integration/Framework/DSL.hs:2594]:
```suggestion
listTransactionsWithLimit ctx w limit = do
```

File: lib/wallet/test/unit/Cardano/WalletSpec.hs [hunk 4]
```diff
@@ -530,6 +533,21 @@ walletListTransactionsSorted wallet@(wid, _, _) _order (_mstart, _mend) history
                 | (tx, meta) <- history ]
         times `shouldBe` expTimes
 
+walletListTransactionsLimited
+    :: (WalletId, WalletName, DummyState)
+    -> SortOrder
+    -> (Maybe UniformTime, Maybe UniformTime)
+    -> [(Tx, TxMeta)]
+    -> Property
+walletListTransactionsLimited wallet@(wid, _, _) _order (_mstart, _mend) history =
```
Comment (jonathanknowles) [inline lib/wallet/test/unit/Cardano/WalletSpec.hs]:
```suggestion
walletListTransactionsWithLimit wallet@(wid, _, _) _order (_mstart, _mend) history =
```

File: lib/wallet/api/http/Cardano/CLI.hs [hunk 5]
```diff
@@ -916,6 +916,7 @@ data TransactionListArgs = TransactionListArgs
     , _timeRangeEnd :: Maybe Iso8601Time
     , _sortOrder :: Maybe SortOrder
     , _schema :: TxMetadataSchema
+    , _limit :: Maybe Int
```
Comment (jonathanknowles) [inline lib/wallet/api/http/Cardano/CLI.hs]: Question: is the behaviour defined for negative values?

If there's no reason to support negative values, then perhaps we could use `Natural` or `Word` here?

File: specifications/api/swagger.yaml [hunk 6]
```diff
@@ -3992,6 +3992,14 @@ x-parametersSortOrder: &parametersSortOrder
       - descending
     default: descending
 
+x-parametersMaxCount: &parametersMaxCount
+  in: query
+  name: max_count
+  description: |
+    An optional maximum count.
+  schema:
+    type: integer
```
Comment (jonathanknowles) [inline specifications/api/swagger.yaml:4001]: Suggestion: specify a minimum, either `0` or `1`. (Though not sure if there's much value in being able to specify `0`.)

```suggestion
    type: integer
    minimum: 1
```

File: lib/wallet/integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs (hunk 2 above)
Comment (paolino) [reply lib/wallet/integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs:1859]: covered in unit tests

File: lib/wallet/api/http/Cardano/CLI.hs (hunk 5 above)
Comment (paolino) [reply lib/wallet/api/http/Cardano/CLI.hs]: No we are not offering a negative number of results, this is just a wallet :-)

**Comment by piotr-iohk** [comment]
bors try

**Review by piotr-iohk — APPROVED** [review]
LGTM. I've added a bit more integration tests and benchmark for listTx = 100 to restorationBenchmarks suite such that we can control the expectation that `max_count` makes it faster.

**Comment by paolino** [comment]
bors retry

**Pull request merged**